### ISSUE TITLE
OSX Catalina needs additional permissions?

### ISSUE THREAD

**Issue by s-andrews** [s-andrews/SeqMonk#219]
We've had a few reports in from OSX users with odd looking failures in straight forward disk access.

For example:

```
SeqMonk Version:1.45.4

Operating System:Mac OS X - 10.15.2

Java Version:11 - 11+28

java.lang.IllegalStateException
java.io.IOException: Operation not permitted

uk.ac.babraham.SeqMonk.DataTypes.DataSet$ChromosomeDataStore.run(DataSet.java:874)
java.base/java.lang.Thread.run(Unknown Source)

```

or

```
SeqMonk Version:1.45.4

Operating System:Mac OS X - 10.15.2

Java Version:11 - 11+28

java.io.FileNotFoundException
/Volumes/DRIVE/SeqData/file.smk (Operation not permitted)

java.base/java.io.FileInputStream.open0(Native Method) java.base/java.io.FileInputStream.open(Unknown Source) java.base/java.io.FileInputStream.<init>(Unknown Source) java.base/java.io.FileReader.<init>(Unknown Source)
uk.ac.babraham.SeqMonk.DataParsers.SeqMonkParser.parseFile(SeqMonkParser.java:150)

```

I think this relates to:

https://www.macworld.com/article/3311982/the-difference-between-accessibility-and-full-disk-access.html

where the new version of OSX won't let apps access data on removable drives without having an additional privilege added in control panel.  I need to find a way to trigger this here so I can test this.

**Comment by s-andrews** [comment]
This might also be relevant since I suspect that asking for more permissions might require the app to be signed.

https://devreboot.wordpress.com/2014/07/04/distributing-an-app-on-mac-os-x-without-xcode-outside-the-mac-app-store/

**Comment by s-andrews** [comment]
This might be relevant:

https://attack.mitre.org/techniques/T1144/

It suggests that the com.apple.quarantine attribute might be the thing which triggers gatekeeper.  If so then it could be removed with:

```sudo xattr -r -d com.apple.quarantine SeqMonk.app```

..I need to find a system to test this on.

**Comment by s-andrews** [comment]
This article seems to have a manual work round for the file reading limitations:

https://nektony.com/duplicate-finder-free/folders-permission

Specifically, this section describes how to allow access to different drive areas for specific apps:

https://nektony.com/duplicate-finder-free/folders-permission#folders

It would be really good if someone could test these instructions and see if that actually works.

**Comment by s-andrews** [comment]
It looks like the option in the last note might be the way to fix this.  I tried doing this on my existing Mavericks system, which doesn't have the "Files and Folders" option, but does have "Full Disk Access".  I tried adding seqmonk to the permitted apps list, but when I add it it says "There was an error in Security & Privacy preferences" - not sure what that means.

I'd be really interested if any Catalina users could check whether adding seqmonk to the "Files and Folders" preference fixes things for them.

**Comment by s-andrews** [comment]
More potentially useful information.  It seems that the system doesn't like scripts which spawn other scripts, only binaries:

https://n8henrie.com/2018/11/how-to-give-full-disk-access-to-a-binary-in-macos-mojave/

That might mean we'd need to add perl and possibly java to the allowed list which doesn't sound like a great solution.  If I could get away with adding the embedded java to the permissions that would work, but the dialog gives me no way to access the contents of an app bundle to add this permission.

**Comment by TimVMoser** [comment]
Ahoy, 
I have just tested the folder-permission workaround under MAC OS Catalina. SeqMonk does not show up in the list of Software that I can choose to give permission to and the "plus" allowing me to add software to the list is greyed out even after unlocking changes. However, I was able to add SeqMonk to the list of Software with full disk access which in turn adds it to the list of software with access to files and folder. Unfortunately to no avail.  
I jcame across this thread discussing a similar issue experienced with geany. I do not know if the two cases are actually related but if they are maybe this helps:
https://github.com/geany/geany/issues/2344

Thank you for the great support, really quite amazing for an open source tool.

Best Regards 
Tim

**Comment by TimVMoser** [comment]
I found a rather lame workaround. As it turns out SeqMonk has full permission for /Users/"your_username"/  which is also the location where the SeqMonk genomes and cache folders are located. I Added a folder to this location (/Users/"my_username"/seqmonk_import 
) and as long as I copy all my input files there I encounter no issues whatsoever.
Cheers

**Comment by s-andrews** [comment]
> 
> 
> I found a rather lame workaround. As it turns out SeqMonk has full permission for /Users/"your_username"/ which is also the location where the SeqMonk genomes and cache folders are located. I Added a folder to this location (/Users/"my_username"/seqmonk_import
> ) and as long as I copy all my input files there I encounter no issues whatsoever.
> Cheers

Yes, I know that home directories are allowed by default, so at least it's not completely broken, but I really want to get externally mounted drives going too - that will be the most common route by which people get data into a project.

**Comment by s-andrews** [comment]
> I have just tested the folder-permission workaround under MAC OS Catalina. SeqMonk does not show up in the list of Software that I can choose to give permission to and the "plus" allowing me to add software to the list is greyed out even after unlocking changes. However, I was able to add SeqMonk to the list of Software with full disk access which in turn adds it to the list of software with access to files and folder. Unfortunately to no avail.

OK, so you got one step further than me as I got an error when I tried adding it to full disk access.

I think the reason it's failing is that it uses a script to launch the main program, so the 'binary' which is being launched is ```/usr/bin/perl``` so that's probably what needs the permissions in the current setup.

This isn't going to be the proper fix eventually, but can you see if you can add ```/usr/bin/perl``` to to full disk access permissions, and see if that makes things work (it might not as the script then spawns a java process so that might also need to be allowed!).  

In the end it looks like I might need to write a binary launch script to have something I can give permissions to - I already do this for Windows, so maybe we can increase the number of languages in SeqMonk by adding something like Rust to Perl, Java, C# and R!

**Comment by TimVMoser** [comment]
Ahoy,

Additionally adding /usr/bin/perl actually did the trick!
 I cannot detect anything missing from the file-browser import pop up, at least not in any of the locations I have checked. No other permission request or anything showed either. I am quite happy with this.
However /usr is a hidden folder and does not show up when using the window that appears after clicking the little plus  (System Pref/Sec.&Priv./Full Disk Acc. > little plus>Window). 
But, since hidden files are displayed in my normal finder windows I just pulled it into the full disc access list. 
Cheers Tim

**Comment by s-andrews** [comment]
Great - thanks for confirming @TimVMoser !  Your help is really appreciated.

At least I now have a fix I can offer to people who face this issue immediately, but it won't be the eventual solution for this issue.  I think the right answer will be to write a binary launch program which can have permissions added to it so that we can add the seqmonk app directly.

I guess I need to go read up on objectiveC / Rust / C++ then...

**Comment by s-andrews** [comment]
I've now written a (very) minimial binary launcher in Rust which compiles to a native Mach executable so it should be suitable to receive permissions.  This was added in commits 10c2cff 1ba816c and ddb2d0d 

However when I now try to add the SeqMonk.app to the Full Disk Access list nothing happens.  It doesn't get added, but no messages warnings errors, nothing.  This is all on Mojave so maybe it works better or at least gives an informative error on Catalina.  Is anyone able to test it there?

I've put the updated binary at https://www.bioinformatics.babraham.ac.uk/projects/seqmonk/seqmonk_v1.46.1.devel_osx64.dmg

**Comment by TimVMoser** [comment]
Ahoy,
Thanks for putting in the effort, unfortunately I have no access to my office Mac for the foreseable future (minimum 2 weeks).

I am happy to test it as soon as we are allowed back to work.

Cheers

________________________________

**Comment by s-andrews** [comment]
Doing some digging on the Mac forums it seems that it might be a requirement to have the app in the Applications folder in order to add additional permissions so that might (partially) explain why it's still not working.  I'll try to find a machine to do some testing.

**Comment by s-andrews** [comment]
I just tried putting seqmonk into my applications folder and once I'd done that it magically appeared in the full disk access category.  Hopefully this means that it will pick up the permissions correctly.  I'll try to find a test I can run to verify this.

![image](https://user-images.githubusercontent.com/8157923/81295556-81637980-9068-11ea-8d58-e63cad449ba7.png)

**Comment by s-andrews** [comment]
I did a bit more testing and there are still problems.  The Rust launcher is currrently broken in that the working directory isn't correctly set, which means that I can't properly test the app bundle to see if the new permissions are actually working.  I think I know how to fix this so I'll need to update the Rust code and rebuild.

**Comment by s-andrews** [comment]
Touch wood, but I think this is fixed!!!

In the end it was a combination of writing a native OSX binary launcher in Rust and also putting SeqMonk into the Applications folder which then allows you to add the necessary permissions for full disk access.

I've also added an OSX specific check to the welcome screen so it will check if you have full access already, and if not it will point you to a new help page describing how to fix it.

I've updated the development snapshot at https://www.bioinformatics.babraham.ac.uk/projects/seqmonk/seqmonk_v1.46.1.devel_osx64.dmg with a version which supports disk access.  I still don't have access to a Catalina machine so I've only tested it on 10.14 but I'm reasonably confident that it will work the same way on 10.15.

If anyone can test the new snapshot that would be super helpful, and if it works then I can put this out as a new release.
